---
title: "Gärtner et al. Plasmacytoid dendritic cells regulate tissue homeostasis of megakaryocytes"
author: "Visha"
date: "13/10/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
#Gärtner et al. Plasmacytoid dendritic cells regulate tissue homeostasis of megakaryocytes

```{r reading_in_data, echo=TRUE}
library(SoupX)
library(dplyr)
library(patchwork)
library(sctransform)
library(ggplot2)
library(dittoSeq)
library(RColorBrewer)
library(data.table)
library(magrittr)
library(readxl)
library(reshape2)
library(enrichR)
library(Seurat)
library(clusterProfiler)
library(stringr)
library(EnhancedVolcano)
library(dittoSeq)
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)


```

#### The Seurat object utilized in this analysis is named MKP3HTO. The conditions are named as follows: Platelet depletion (PD), Bl6/control (Bl6), platelet depletion with pDC depletion (PD+pDC)
#### This code block performs a series of data analysis and visualization steps on single-cell RNA sequencing data using the Seurat package in R. The primary focus is on loading a Seurat object, visualizing clusters, identifying marker genes, and plotting these findings. Here's a breakdown of the steps:

#Reading a Seurat Object: The Seurat object MKP3HTO is loaded from an RDS file stored locally. RDS is a format used in R for saving and loading objects.
#Setting the Default Assay: The default assay of the Seurat object is set to 'RNA', which specifies the type of data (RNA sequencing data) to be used in subsequent analyses.
#UMAP Visualization for Clusters: A UMAP (Uniform Manifold Approximation and Projection) plot is created to visualize data clusters based on 'HTO_classification', a specific type of cell classification. UMAP is a dimensionality reduction technique often used in single-cell data analysis.
#Shuffled UMAP Plot for Clustering Robustness: A shuffled UMAP plot is generated, likely to assess the robustness of the clustering. Shuffling the data can help in understanding how well the clusters are defined.
#UMAP Plot with Seurat Clusters: Another UMAP plot is created, this time showing clusters identified by Seurat's clustering algorithm, with each cluster labeled.
#Identifying Marker Genes: The script identifies marker genes in each cluster using statistical methods (here, Wilcoxon test). Marker genes are those significantly expressed in specific clusters and can help in characterizing cell types.
#Selecting Top Marker Genes: The top two marker genes per cluster are selected based on log fold change, highlighting the most differentially expressed genes in each cluster.
#Preparation for Dot Plot: The script prepares data for the top 10 marker genes per cluster for a dot plot visualization. This involves selecting genes based on their average log2 fold change.
#Dot Plot for Marker Genes: Finally, a dot plot is created to visualize the expression of these top 10 marker genes across the different clusters. This type of plot is useful for identifying genes that define each cluster's characteristics.

```{r cluster_marker_genes, echo=TRUE}
# Reading the Seurat object from an RDS file
MKP3HTO = readRDS("~/Downloads/final_MKP3HTO.RDS") # Replace with your file path

# Setting the default assay to 'RNA' for the Seurat object
DefaultAssay(MKP3HTO) <- "RNA"

# Creating a UMAP plot to visualize clusters based on HTO classification
print(DimPlot(MKP3HTO, group.by = "HTO_classification"))

# Indicating a figure label for manuscript reference
print("Manuscript Extended Data Figure 10 a")

# Creating a shuffled UMAP plot for robustness check of clustering
print(DimPlot(MKP3HTO, shuffle = T, seed = 1, group.by= "HTO_classification", split.by= "HTO_classification", ncol=3, raster=FALSE))

# Indicating another figure label for manuscript reference
print("Manuscript Figure 4D")

# Creating a UMAP plot showing different Seurat clusters with labels
print(DimPlot(MKP3HTO, group.by= "seurat_clusters", raster=FALSE, label = TRUE))

# Finding all significant marker genes in the Seurat object
MKP3HTO_markers <- FindAllMarkers(MKP3HTO, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, test.use = "wilcox")

# Selecting the top 2 marker genes per cluster based on log fold change
MKP3HTO_markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

# Preparing data for the top 10 marker genes per cluster for plotting
MKP3HTO_markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10

# Indicating a dot plot figure label for manuscript reference
print("Dotplot of cluster defining genes")

# Indicating another figure label for manuscript reference
print("Manuscript Extended Data Figure 10 b")

# Creating a dot plot to visualize the top 10 marker genes in clusters
print(DotPlot(MKP3HTO, features = unique(top10$gene))+ RotatedAxis())
```

#### dotplot of genes per cluster to determine the clusters more accurately.This code block is designed to create a dot plot visualization of specific gene expression in a Seurat object. It is primarily focused on displaying the expression levels of various gene sets across different cell types or clusters identified in the Seurat object. The genes are categorized into several groups based on their functions or characteristics, such as transcription factors, signaling molecules, surface proteins, and cell cycle-related genes.

```{r dotplot, echo=TRUE}
# Load the required Seurat library

# Load your Seurat object (replace 'seurat_obj' with your Seurat object)
seurat_obj <- MKP3HTO

# Define the gene sets as a list of lists
# Each list contains genes related to specific cell functions or types
MK_TF = c("Cited2", "Fli1", "Pbx1", "Mef2c", "Meis1") # Megakaryocyte transcription factors
MK_GRANULES = c("Pf4", "Vwf", "Clu", "Ppbp", "Serpine2", "Thbs1", "F5", "Ctla2a", "Angpt1") # Megakaryocyte granules
MK_SIGNALING = c("Rap1b", "Cavin2", "Tpm4", "Myl9", "Gng11", "Alox12", "Tmsb4x", "Plek", "Prkg1", "Prkca", "Trpc6") # Megakaryocyte signaling molecules
MK_SURFACE = c("Itga2b", "Mpl", "Cd9", "Gp9", "Gp1bb", "Gp5", "Gp6", "Clec1b", "Slamf1") # Megakaryocyte surface proteins
ERY_TF = c("Zfpm1", "Hmbg3", "Gata1", "Gfi1b", "Klf1") # Erythroid transcription factors
ERY = c("Eng", "Car2", "Hba-a2", "Gypa") # Erythroid-specific genes
GMP_TF = c("Gata2", "Cebpa", "Irf8") # Granulocyte-monocyte progenitor transcription factors
GMP = c("Elane", "Mpo") # Granulocyte-monocyte progenitor-specific genes
NEUTRO_TF = c("Spi1", "Cebpe") # Neutrophil transcription factors
NEUTRO = c("Camp", "Ltf", "Retnlg") # Neutrophil-specific genes
NK_TF = c("Ets1", "Foxo1", "Runx3") # Natural killer cell transcription factors
NK = c("Skap1", "Ccl5", "Id2") # Natural killer cell-specific genes
BASO_TF = c("Lmo4", "Runx1", "Stat5b") # Basophil transcription factors
BASO = c("Hdc", "Cd200r3", "Prss34") # Basophil-specific genes
Cell.cycle = c("Nusap1", "Top2a", "Mki67") # Cell cycle-related genes
MIXED_RB = c("Ncl", "Npm1", "Cmss1") # Mixed group (ribosome biogenesis)
MIXED_TI = c("Eif4b", "Eif4a1", "Eif2s2", "Pdcd4") # Mixed group (translation initiation)
MIXED_NM = c("Nme1", "Dctpp1") # Mixed group (nucleotide metabolism)
MIXED_PC = c("Hsp90aa1", "Hsp90ab1") # Mixed group (protein chaperones)

# Combine all gene sets into one list for plotting
features <- list("MK TF" = MK_TF, "MK_GRANULES" = MK_GRANULES, "MK_SIGNALING" = MK_SIGNALING, "MK_SURFACE" = MK_SURFACE, 
                 "ERY_TF" = ERY_TF, "GMP_TF" = GMP_TF,"GMP"=GMP,"NEUTRO_TF" = NEUTRO_TF,"NEUTRO" = NEUTRO, 
                 "NK_TF"=NK_TF,"NK" = NK,"BASO_TF" = BASO_TF,"BASO" = BASO, "Cell.cycle" = Cell.cycle,
                 "MIXED_RB"=MIXED_RB,"MIXED_TI" = MIXED_TI, "MIXED_NM" = MIXED_NM, "MIXED_PC" = MIXED_PC)

# Indicate figure title for manuscript reference
print("Manuscript Figure 4E")

# Create and print the dot plot using the defined features
d = DotPlot(object = MKP3HTO, features=features, cluster.idents=T) + theme(axis.text.x = element_text(angle = 90))
print(d)

```

#### This code block is designed to generate a bar plot visualizing the proportion of cells in each cluster for different conditions in a Seurat object named MKP3HTO. The plot aims to show how cell composition varies across clusters identified by a resolution of 0.25 in the Seurat object, with a focus on a variable named "CSclassification". This type of visualization is useful for understanding the distribution of different cell types or conditions across identified clusters in single-cell RNA sequencing data.
```{r proportion, echo=TRUE}

print("getting proportion of cells per cluster for each condition")

# Indicating the figure title for manuscript reference
print("Manuscript Figure 4E")

# Generating a bar plot showing the proportion of cells per cluster
# 'object': specifies the Seurat object (MKP3HTO) to be used
# 'var': the variable ('CSclassification') used to group the cells
# 'group.by': clustering resolution, here at 0.25, to categorize cells
dittoBarPlot(
    object = MKP3HTO,
    var = "CSclassification",
    group.by = "RNA_snn_res.0.25")

```


#### This code block is designed to project and analyze differentially expressed (DE) genes from bulk RNA sequencing onto single-cell clusters using a Seurat object (MKP3HTO). The analysis focuses on genes upregulated in PD compared to Control (CNTRL) and downregulated in PD combined in condition PDC+PD compared to CNTRL. The AddModuleScore function is used to score these genes in the single-cell data, and the results are visualized using violin plots to understand the gene expression patterns across different cell states or types.
```{r bulk_seq_de_genes, echo=TRUE}
# Reading the bulk sequencing differentially expressed genes from a CSV file
bulk_de_genes = read.csv("~/Downloads/MKP_samples_log2FC_forVisha.csv", header = TRUE)

# Setting the gene names as row names for the data frame
rownames(bulk_de_genes) = bulk_de_genes$Gene_name

# Selecting genes upregulated in PD vs CNTRL
genes_pd_vs_cntrl  <- subset(bulk_de_genes, log2FC_PD_vs_Ctrl > 0)
genes_pd_vs_cntrl = rownames(genes_pd_vs_cntrl)

# Selecting genes downregulated in PDC+PD vs CNTRL
genes_pdpdc_vs_pd = subset(bulk_de_genes, log2FC_PD_DT_vs_PD < 0)
genes_pdpdc_vs_pd = rownames(genes_pdpdc_vs_pd)

# Finding the intersection of upregulated and downregulated genes
final_genes = intersect(genes_pd_vs_cntrl, genes_pdpdc_vs_pd)

# Adding a module score for these genes in the Seurat object
MKP3HTO <- AddModuleScore(MKP3HTO, features = list(final_genes), name = "final_genes")

# Subsetting the Seurat object to exclude certain classifications
MKP3HTO = subset(x = MKP3HTO, subset = CSclassification=="pDC-depletion", invert = TRUE)

# Setting the identity classes for the Seurat object
Idents(MKP3HTO) = MKP3HTO$Idents

# Creating and printing violin plots to visualize the module scores
# Different plots show the distribution of scores across various classifications and identities
print(VlnPlot(MKP3HTO, features = "final_genes1", group.by ="CSclassification"))
print(VlnPlot(MKP3HTO, features = "final_genes1", split.by ="CSclassification"))
print(VlnPlot(MKP3HTO, features = "final_genes1", group.by ="CSclassification", split.by = "Idents"))
print(VlnPlot(MKP3HTO, features = "final_genes1", group.by = "Idents"))


```


##### This code block focuses on identifying differentially expressed (DE) genes in a specific cluster of a Seurat object (MKP3HTO). The cluster of interest is 'MK-MEP_cycling', represented by seurat_clusters == 3. The analysis is performed for two sets of conditions: PD vs Bl6 and PD vs pDC-PD. The Wilcoxon test is used to find DE genes, with selection criteria being an average log fold change (avg_log2FC) greater than 0.25 and an adjusted p-value (p_val_adj) less than 0.05. The identified DE genes are then visualized using volcano plots with the help of the EnhancedVolcano package. Volcano plots are effective for visualizing significant changes in gene expression between two conditions, highlighting genes that are notably upregulated or downregulated.
```{r de_results, echo=TRUE}
# Subsetting the Seurat object for cluster 3 (MK-MEP_cycling)
subset_obj = subset(x = MKP3HTO, subset = seurat_clusters == 3)

# Setting the identities in the subset object based on CSclassification
Idents(subset_obj) = subset_obj$CSclassification

# Extracting gene names for the cluster of interest
gene_names <- rownames(subset_obj@assays$RNA@scale.data)[subset_obj$seurat_clusters == 3]

# Finding DE genes for PD vs Bl6 condition using Wilcoxon test
de_results <- FindMarkers(
      subset_obj,
      ident.1 = "Platelet-depletion",
      ident.2 = "Bl6",
      test.use = "wilcox",
      slot= "scale.data",
      genes.use = gene_names  # Using specified gene names for comparison
    )

# Preparing a volcano plot for visualizing DE results
print("Manuscript Figure 4h")
volcano = EnhancedVolcano(de_results,
    lab = rownames(de_results),
    x = 'avg_diff',
    y = 'p_val_adj',
    title = "PD vs CNTRL in cycling_MK_MEP cluster",
    pCutoff = 0.05,  # P-value cutoff for significance
    FCcutoff = 0.25, # Fold change cutoff for significance
    # Additional plot parameters for aesthetics and readability
    pointSize = 3.0,
    colAlpha = 1,
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.2,
    max.overlaps = 30,
    colConnectors = 'grey30',
    labSize = 6.0)
print(volcano)
# Repeat the process for PD vs pDC-Platelet-depletion condition
de_results <- FindMarkers(
      subset_obj,
      ident.1 = "Platelet-depletion",
      ident.2 = "pDC-Platelet-depletion",
      test.use = "wilcox",
      slot= "scale.data",
      genes.use = gene_names  # Using specified gene names for comparison
    )

# Preparing another volcano plot for the second condition comparison
print("Manuscript Figure 4i")
volcano = EnhancedVolcano(de_results,
    lab = rownames(de_results),
    # Parameters similar to the previous volcano plot
    x = 'avg_diff',
    y = 'p_val_adj',
    title = "PD vs CNTRL in cycling_MK_MEP cluster",
    pCutoff = 0.05,
    FCcutoff = 0.25,
    # Additional plot parameters
    pointSize = 3.0,
    colAlpha = 1,
    legendLabSize = 12,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 0.2,
    max.overlaps = 30,
    colConnectors = 'grey30',
    labSize = 6.0)
print(volcano)
```

## reading in the top 40 response to type I interferon genes from the bulk-seq reactome pathway analysis 
```{r interferon, echo=TRUE}
DefaultAssay(MKP3HTO) <- "RNA"
## response to type I interferon genes as bulk_seq genes
bulk_seq = read.csv("~/Downloads/GOBP_RESPONSE_TO_TYPE_I_INTERFERON.csv", header = TRUE)
print("response to interferon genes")
head(bulk_seq)


## getting the top 40 genes from response to type I interferon genes
bulk_seq = as.data.frame(bulk_seq)
rownames(bulk_seq) = bulk_seq$PROBE
bulk_seq_genes = head(rownames(bulk_seq), 40)
## converting the genes to mouse gene form - "Tyk2" "Hla-a"
bulk_seq_genes = tolower(bulk_seq_genes)
firstup = function(x) {
  substr(x, 1, 1) = toupper(substr(x, 1, 1))
  
  x
}
bulk_seq_genes = firstup(bulk_seq_genes)
print("response to interferon genes after processing")
bulk_seq_genes
### renaming the idents of seurat obj to conditions instead of clusters to calculate fold change between conditions for the genes
Idents(MKP3HTO) = MKP3HTO$CSclassification


```

# scatterplots to plot the common DE genes between single-cell clusters (MK-MEP cycling and MK-MEP) and bulk-seq for the comparison PD vs CNTRL. the axis' represent the expression levels of the genes in the experiments.
```{r scatter_plots, echo=TRUE}


subset_obj = subset(x = MKP3HTO, subset = seurat_clusters == 3)
Idents(subset_obj) = subset_obj$CSclassification
# finding the DE results for all genes of the seurat object
gene_names <- rownames(subset_obj@assays$RNA@scale.data)[subset_obj$seurat_clusters == 3]
de_results <- FindMarkers(
      subset_obj,
      ident.1 = "Platelet-depletion",
      ident.2 = "Bl6",
      test.use = "wilcox",
      slot= "scale.data",
      genes.use = gene_names  # Specify the gene names for the comparison
    )
# getting the interferon response genes
interferon = read.csv("~/Downloads/GOBP_RESPONSE_TO_TYPE_I_INTERFERON.csv", header = TRUE)
## converting the genes to mouse gene form - "Tyk2" "Hla-a"
interferon$PROBE = tolower(interferon$PROBE)
firstup = function(x) {
  substr(x, 1, 1) = toupper(substr(x, 1, 1))
  
  x
}
# getting the top 40 interferon response genes based on the heatmap of bulk-seq results
interferon$PROBE = firstup(interferon$PROBE)
rownames(interferon) = interferon$PROBE
interferon = head(rownames(interferon), 40)
# The interferon genes are
print(interferon)

# subsetting the de_results data frame to only get the interferon genes
de_results_inf = na.omit(de_results[interferon, ])
de_results_inf$Gene = rownames(de_results_inf)
# adding a column called FDR which adjustes p val for all the interferon genes
de_results_inf$FDR = p.adjust(de_results_inf[rownames(de_results_inf), 'p_val'])

# getting the bulk_seq results file 
bulk_de_genes = read.csv("~/Downloads/MKP_samples_log2FC_forVisha.csv", header = TRUE)
rownames(bulk_de_genes) = bulk_de_genes$Gene_name
genes_pd_vs_cntrl_pos  <- bulk_de_genes
# subsetting the bulk-seq results to only focus on PD vs Cntrl columns
genes_pd_vs_cntrl_pos = genes_pd_vs_cntrl_pos[, c("log2FC_PD_vs_Ctrl", "FDR_PD_vs_control")]
# subsetting the interferon genes and their corresponding FC and FDR from bulk-seq results
genes_pd_vs_cntrl_pos = genes_pd_vs_cntrl_pos[rownames(de_results_inf), ]
genes_pd_vs_cntrl_pos = na.omit(genes_pd_vs_cntrl_pos)
genes_pd_vs_cntrl_pos$Gene = rownames(genes_pd_vs_cntrl_pos)


# merging the interferon fc and fdr values to the bulk-seq fc and fdr values for the condition PD vs CNTRL based on the common interferon genes (36 of them)
merged_data <- na.omit(merge(de_results_inf, genes_pd_vs_cntrl_pos, by = "Gene", all = TRUE))
rownames(merged_data) = merged_data$Gene


print("Manuscript Extended Data Figure 10 e")
# Create the scatterplot for the interferon genes. Plotting FC of scRNAseq on the x-axis and the FC of bulk-seq on the y-axis and colouring based on FDR calculated above using p.adjust
p <- ggplot(data = merged_data, aes(x = avg_diff, y = log2FC_PD_vs_Ctrl, label = Gene, color = FDR)) +
  geom_point() +
  geom_text(hjust = 0.001, vjust = 0.001, nudge_x = 0.001, nudge_y = 0.001, size = 3) +
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "Single Cell Cluster MK-MEP_cycling", y = "Bulk Seq PD vs Bl6") +
  ggtitle("Scatterplot of Fold Changes for interferon response genes PD vs Bl6") +
  ylim(min(merged_data$log2FC_PD_vs_Ctrl) * 0.9, max(merged_data$log2FC_PD_vs_Ctrl) * 1.1) + # Adjusting y-axis limits
  xlim(min(merged_data$avg_diff) * 0.9, max(merged_data$avg_diff) * 1.1) # Adjusting x-axis limits

print(p)


######################################################################################


# getting the common DE genes from cluster 3 PD vs CNTL and bulk-seq PD vs CNTRL from already calculated file
cluster_3_Platelet_depletion_vs_Bl6_common_genes_pos = read.csv("~/Downloads/cluster_3_Platelet-depletion_vs_Bl6_common_genes_pos.csv")
cluster_3_Platelet_depletion_vs_Bl6_common_genes_pos = cluster_3_Platelet_depletion_vs_Bl6_common_genes_pos$x
# subsetting the de_results to only contain the common DE genes
common = na.omit(de_results[cluster_3_Platelet_depletion_vs_Bl6_common_genes_pos, ])
common$Gene = rownames(common)
# calculating the FDR for common genes
common$FDR = p.adjust(common[rownames(common), 'p_val'])

genes_pd_vs_cntrl_pos  <- bulk_de_genes
# subsetting the bulk-seq results to only focus on PD vs Cntrl columns
genes_pd_vs_cntrl_pos = genes_pd_vs_cntrl_pos[, c("log2FC_PD_vs_Ctrl", "FDR_PD_vs_control")]
# subsetting the common DE genes and their corresponding FC and FDR from bulk-seq results
genes_pd_vs_cntrl_pos = genes_pd_vs_cntrl_pos[cluster_3_Platelet_depletion_vs_Bl6_common_genes_pos, ]
genes_pd_vs_cntrl_pos = na.omit(genes_pd_vs_cntrl_pos)
genes_pd_vs_cntrl_pos$Gene = rownames(genes_pd_vs_cntrl_pos)

# merging the common DE genes' fc and fdr values to the bulk-seq fc and fdr values for the condition PD vs CNTRL based on the common interferon genes (61 of them)
merged_data <- merge(common, genes_pd_vs_cntrl_pos, by = "Gene", all = TRUE)
rownames(merged_data) = merged_data$Gene
merged_data = na.omit(merged_data)

print("Manuscript Extended Data Figure 10 d")
# Create the scatterplot for the =common DE genes. Plotting FC of scRNAseq on the x-axis and the FC of bulk-seq on the y-axis and colouring based on FDR calculated above using p.adjust
p <- ggplot(data = merged_data, aes(x = avg_diff, y = log2FC_PD_vs_Ctrl, label = Gene, color = FDR)) +
  geom_point() +
  geom_text(hjust = 0.001, vjust = 0.001, nudge_x = 0.001, nudge_y = 0.001, size = 3) +
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "Single Cell Cluster MK-MEP_cycling", y = "Bulk Seq PD vs Bl6") +
  ggtitle("Scatterplot of Fold Changes for interferon response genes PD vs Bl6") +
  ylim(min(merged_data$log2FC_PD_vs_Ctrl) * 0.9, max(merged_data$log2FC_PD_vs_Ctrl) * 1.1) + # Adjusting y-axis limits
  xlim(min(merged_data$avg_diff) * 0.9, max(merged_data$avg_diff) * 1.1) # Adjusting x-axis limits

print(p)

# enrichment results of the common genes between bulk-seq and single-cell MK-MEP_cycling cluster for PD vs Bl6 conditions.
enrichment_results <- enrichR::enrichr(
              genes = cluster_3_Platelet_depletion_vs_Bl6_common_genes_pos,
              databases = "GO_Biological_Process_2023"
            )
            enrichment_results = as.data.frame(enrichment_results)
                enrichment_data_pos <- data.frame(
          Term = enrichment_results$GO_Biological_Process_2023.Term,
          Overlap = enrichment_results$GO_Biological_Process_2023.Overlap,
          P.value = enrichment_results$GO_Biological_Process_2023.P.value,
          Adjusted.P.value = enrichment_results$GO_Biological_Process_2023.Adjusted.P.value,
          Odds.Ratio = enrichment_results$GO_Biological_Process_2023.Odds.Ratio,
          Combined.Score = enrichment_results$GO_Biological_Process_2023.Combined.Score,
          Genes = enrichment_results$GO_Biological_Process_2023.Genes
        )

print("Manuscript Extended Data Figure 10 d")
enrich_pos = plotEnrich(enrichment_data_pos)
print(enrich_pos)                  
          
```










